### PR TITLE
Add lint recommending using `std::iter::once` and `std::iter::empty` 

### LINKED ISSUE #9186: Lints recommeding to use `std::iter::once` and `std::iter::empty`
### What it does

It seems suitable to separate it in two lints, one for `iter::once` and one for `iter::empty`. They'd look at patterns such as `[foo].into_iter()` and `[].into_iter()` and suggest using `once` and  `empty` instead.

### Lint Name

iter_once and iter_empty

### Category

pedantic

### Advantage

`once` and `empty` take less memory than their the arrays or slice iterators for the corresponding use case.

### Drawbacks

Depending on the use case, the fix can cause type issues, for example if the iterator is then passed to a function that expects an `std::slice::Iter` and not a genreric `impl Iterator`

### Example

```rust
let a = [foo].iter();
let b = [bar].iter_mut();
let c = [baz].into_iter();
let d = Some(foo).iter();
let e = Some(bar).iter_mut();
let f = Some(baz).into_iter();
```
Could be written as:
```rust
use std::iter;
let a = iter::once(&foo);
let b = iter::once(&mut bar);
let c = iter::once(bar);
let d = iter::once(&foo);
let e = iter::once(&mut bar);
let f = iter::once(bar);
```

and 

```rust
let a = [].iter();
let b = [].iter_mut();
let c = [].into_iter();
let d = None.iter();
let e = None.iter_mut();
let f = None.into_iter();
```

could be written as:

```rust
use std::iter;
let a = iter::empty();
let b = iter::empty();
let c = iter::empty();
let d = iter::empty();
let e = iter::empty();
let f = iter::empty();
```

### PR DESCRIPTION
**Pull request by sgued**
```
changelog: [`iter_once`]: add new lint
changelog: [`iter_empty`]: add new lint
```

fixes #9186 

- \[ ] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

The lint doesn't really follow the naming conventions. I don't have any better idea so I'm open to suggestions.

### CODE REVIEW

File: clippy_lints/src/iter_once_empty.rs
Comment (Serial-ATA) [inline clippy_lints/src/iter_once_empty.rs:68]: This should also mention `iter::once`.

File: clippy_lints/src/iter_once_empty.rs
Comment (Serial-ATA) [inline clippy_lints/src/iter_once_empty.rs:79]: You don't need an exhaustive list here, you can just put a few examples.

File: clippy_lints/src/iter_once_empty.rs
Comment (Serial-ATA) [inline clippy_lints/src/iter_once_empty.rs:32]: You don't need an exhaustive list here, you can just put a few examples.

**Review by Serial-ATA — CHANGES_REQUESTED** [review]

File: clippy_lints/src/iter_once_empty.rs
Comment (sgued) [reply clippy_lints/src/iter_once_empty.rs:68]: Why? The two lints are distinct.

File: clippy_lints/src/iter_once_empty.rs
Comment (sgued) [reply clippy_lints/src/iter_once_empty.rs:87]: If the option is not constructed just before calling `iter` this lint shouldn't apply.

File: clippy_lints/src/iter_once_empty.rs
Comment (Serial-ATA) [reply clippy_lints/src/iter_once_empty.rs:60]: Ignore this. That was me being too hasty :sweat_smile:.

File: clippy_lints/src/iter_once_empty.rs
Comment (Serial-ATA) [reply clippy_lints/src/iter_once_empty.rs:87]: Hypothetically, someone could make their own `Some` and `None`. This should only check for `Option`. [Here's an example](https://github.com/rust-lang/rust-clippy/blob/4562dd0a8e45af4fc2a49951df07ce5d3f1108ec/clippy_lints/src/if_then_some_else_none.rs#L77-L79) of how you can check for this.

Comment (sgued) [reply clippy_lints/src/iter_once_empty.rs]: Ok, I changed it to both check the path **and** check the type so we're sure that it's not something else that's also called `Some` for some reason

File: clippy_lints/src/iter_once_empty.rs
Comment (sgued) [reply clippy_lints/src/iter_once_empty.rs]: Done

File: clippy_lints/src/iter_once_empty.rs
Comment (sgued) [reply clippy_lints/src/iter_once_empty.rs]: Ok, I didn't see your message for some reason. So I guess I implemented the right thing

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2345]: Same here.

File: clippy_lints/src/methods/mod.rs
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2367]: I'm a little concerned about this. From how many allows are necessary in Clippy already, I think this is a bigger problem, that has to be addressed before we can merge this.

File: clippy_lints/src/methods/iter_once_empty.rs
Comment (flip1995) [inline clippy_lints/src/methods/iter_once_empty.rs:67]: Applicability is `MaybeIncorrect` for this lint.

File: clippy_lints/src/methods/iter_once_empty.rs
Comment (flip1995) [inline clippy_lints/src/methods/iter_once_empty.rs:66]: The suggestion creation should depend on if it is a `no_std` crate or not. We have other lints that already do that, so you can reuse code from there.

File: clippy_lints/src/methods/iter_once_empty.rs
Comment (flip1995) [inline clippy_lints/src/methods/iter_once_empty.rs:55]: The message should describe the problem, not the solution. This is what the suggestion is for. So this should be something along the lines of "calling `iter` on an empty collection". With this change, all of this match block should collapse into a single `format!` macro.

**Review by flip1995 — CHANGES_REQUESTED** [review]

File: clippy_lints/src/methods/mod.rs
Comment (sgued) [reply clippy_lints/src/methods/mod.rs:2353]: Does clippy have a way to know whether a suggestion type-checks?
Maybe we could limit it to expressions within a `chain`, which would likely lead to less false positives but not 0. It would also create false negatives in cases like:

```rust
let iter = if some_condition {
    [a].into_iter()
} else {
    [b].into_iter()
}
  
iter.chain(other_iter)
```

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: Do you have examples where applying the suggestion does not type check? Looking at all the `allow`s you added to Clippy, I don't really see why you couldn't use `iter::empty` there.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: Hm, if I'd had to guess, I would say that this fails type checking when using `iter::empty` because it can't infer the type anymore? I don't think this is a problem for `iter_once` though.

**Review by flip1995 — COMMENTED** [review]
I'll be thinking about a a lint name before merging this.

File: clippy_lints/src/methods/mod.rs
Comment (sgued) [reply clippy_lints/src/methods/mod.rs:2353]: Say you have something like this:

```rust
let iter = match smth {
    "a" => [1,2].iter(),
    "b" => [3].iter(),
    _ => [].iter(),
};
```

The first branch forces `iter` to be of type `slice::Iter<i32>` but suggestions would be to change it to:

```rust
let iter = match smth {
    "a" => [1,2].iter(),
    "b" => iter::once(&3),
    _ => iter::empty(),
};
```

so now `iter` must be at the same time `Once<i32>`, `slice::Iter<i32>` and `Empty<i32>`

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: Hm, limiting this to iter chains also doesn't solve this, e.g.:

```rust
let iter = match smth {
    "a" => [[1],[2]].iter().flatten(),
    "b" => [[3]].iter().flatten(),
    _ => [].iter().flatten(),
};
    
let iter = match smth {
    "a" => [[1],[2]].iter().flatten(),
    //"b" => [[3]].iter().flatten(),
    "b" => std::iter::once([3]).flatten(),
    _ => std::iter::empty().flatten(),
};
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e6b3994fbecbb6b4ab4d9a369c9328db)

I don't really have a solution to this... I like the idea of the lint, but think that this is too big of a FP to have it as a `pedantic` lint. I'll have to check how often this FP appears in the wild.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: Running it on the top 50 crates, those lints were only emitted on `serde_derive` and all 3 cases were fixable. Maybe I making a mountain out of a molehill here.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: I asked on Zulip about this https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60iter_empty.60.2F.60iter_once.60.20new.20lints.20clippy.239187/near/290350623

Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:2353]: > Running it on the top 50 crates, those lints were only emitted on serde_derive and all 3 cases were fixable. Maybe I making a mountain out of a molehill here.

@flip1995 have you ensured that the lint was enabled when you ran it on those crates, or could it be that they haven't enabled `pedantic` lints?

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2353]: I used `@mikerite`s tool to check those lints. I'm pretty sure they were enabled. I mean one of those triggered on serde_derive.

Comment (sgued) [reply clippy_lints/src/methods/mod.rs:2353]: I ran the tool on the top 200 crates to have a better idea.


The results are:
| Crate                        | Count | False Positive (FP)| Partial FP |
|:-----------------------------|:-----:|:-----:             | :--------: |
| chrono-0.4.19                |     1 |               0     |0  |
| petgraph-0.6.2               |     2 |                   1|  0|
| rayon-1.5.3                  |     1 |                   1|  0|
| serde_derive-1.0.140         |     3 |                   0| 0|
| synstructure-0.12.6          |     1 |                   0| 0 |
| tokio-1.20.0                 |     2 |                   0| 2          |
| unicode-normalization-0.1.21 |     6 |                   6| 1          |
| clippy | 7 | 2 | 0 |

*Partial FP* means the lint is right to trigger but the suggestion doesn't typecheck.

#### [petgraph](https://github.com/petgraph/petgraph/blob/master/src/graphmap.rs#L355=) and  [rayon](https://github.com/rayon-rs/rayon/blob/master/src/vec.rs#L279=) 

False positive, a `slice::Iter(Mut)` is needed

#### `serde_derive` ([1](https://github.com/serde-rs/serde/blob/master/serde_derive/src/bound.rs#L344=) [2](https://github.com/serde-rs/serde/blob/master/serde_derive/src/de.rs#L3001=) [3](https://github.com/serde-rs/serde/blob/master/serde_derive/src/de.rs#L3067=)) and [chrono](https://github.com/chronotope/chrono/blob/main/src/datetime/mod.rs#L622=)

Each time the perfect true positive for this lint

#### [tokio](https://github.com/tokio-rs/tokio/blob/master/tokio/src/net/addr.rs#L57=)

Two "false positive". In the sense that the suggestion doesn't TypeCheck but by changing some type annotations it does. This could have required a breaking change if it hadn't been on some private `impls` though.

#### [unicode-normalization](https://github.com/unicode-rs/unicode-normalization/blob/master/src/lib.rs#L173=) 
6 times the same false positive for type checking, because a trait is implemented specifically for `Option::IntoIter`, but it might be interesting for them to implement the trait also for `iter::Once`. I don't know how the crate is used and whether it would be useful though.
 
#### Clippy itself (in this PR)

The lint is fixed 5 times and ignored twice. There was one "partial false positive" which required manually writing the type of `Empty<ty::subst::GenericArg<'_>>`, which was a bit of a pain.

**Comment by flip1995** [comment]
Thanks for doing the analysis (https://github.com/rust-lang/rust-clippy/pull/9187#discussion_r928122711). I nominated this PR for the meeting this evening to figure out if we want to get this lint into `pedantic` or have it in `nursery` for starters.

The meeting is at 3pm UTC: https://everytimezone.com/s/deeec323

**Comment by flip1995** [comment]
So in the Clippy meeting, we decided to place those lints in `nursery` for now and lift them up to `pedantic` once at least the most obvious FPs are addressed. Jarcho left some instructions on how this could be done [here](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60iter_empty.60.2F.60iter_once.60.20new.20lints.20clippy.239187/near/290379301). However I don't want to block this PR on this.

**Comment by sgued** [comment]
I see on Zulip you proposed `iter_on_empty_collection` for the name of `iter_empty`. I like it. I suggest `iter_on_singleton` for `iter_once`

File: clippy_lints/src/methods/mod.rs
Comment (sgued) [reply clippy_lints/src/methods/mod.rs:2353]: 3b45a28090ea95a6ad0a589a581a1297070521cc based on [this](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60iter_empty.60.2F.60iter_once.60.20new.20lints.20clippy.239187/near/290379301) fixed the FP in petgraph and clippy itself. The Lint still triggers in Tokio and unicode-normalization though.

**Comment by flip1995** [comment]
I think of something different when I think about singletons. How about `iter_on_single_items`?

**Review by flip1995 — APPROVED** [review]
This LGTM now. renaming the lints, please squash some (doesn't have to be all) of the commits and this should be ready to go.

**Comment by flip1995** [comment]
Can you squash the commits a bit further, please? Most of those are trivial/one-line changes, that we don't really need in the Clippy history.

**Comment by sgued** [comment]
Is it squashed enough now or do you want even more?

**Comment by flip1995** [comment]
Commits [f30d7c2](https://github.com/rust-lang/rust-clippy/pull/9187/commits/f30d7c2495b3188bfa0cf8bd6dbbf76494c9845e) and [b247594](https://github.com/rust-lang/rust-clippy/pull/9187/commits/b247594a39cb8e2c3d8d193860eac0eb1c63e1ae) could also be squashed just looking at the commit message. But that's just me being pedantic, so I won't block this PR on that.

Thanks for squashing. That's how I (almost :yum:) like my commit messages.

@bors r+

**Pull request merged**
